atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/NameMetadataDescription.java

```diff
@@ -213,47 +225,51 @@ public static NameMetadataDescription hydrateFromProto(TableMetadataPersistence.
         for (TableMetadataPersistence.NameComponentDescription part : message.getNamePartsList()) {
             list.add(NameComponentDescription.hydrateFromProto(part));
         }
+
         // Call constructor over factory because the list might already contain the hash row component
-        return new NameMetadataDescription(list, message.getHasFirstComponentHash());
+        return new NameMetadataDescription(list, getNumberOfComponentsHashedFromProto(message));
     }
 
-    @Override
-    public String toString() {
-        return "NameMetadataDescription [rowParts=" + rowParts + ", hasFirstComponentHash=" + hasFirstComponentHash
-                + "]";
+    private static int getNumberOfComponentsHashedFromProto(TableMetadataPersistence.NameMetadataDescription message) {
+        int numberOfComponentsHashed;
+        if (message.getHasFirstComponentHash()) {
+            numberOfComponentsHashed = 1;
+        } else {
+            numberOfComponentsHashed = message.getNumberOfComponentsHashed();
+        }
+        return numberOfComponentsHashed;
     }
 
     @Override
-    public int hashCode() {
-        final int prime = 31;
-        int result = 1;
-        result = prime * result + (hasFirstComponentHash ? 1231 : 1237);
-        result = prime * result + ((rowParts == null) ? 0 : rowParts.hashCode());
-        return result;
+    public String toString() {
+        return "NameMetadataDescription [rowParts=" + rowParts + ", numberOfComponentsHashed=" +
+                numberOfComponentsHashed + "]";
     }
 
     @Override
     public boolean equals(Object obj) {
         if (this == obj) {
             return true;
         }
-        if (obj == null) {
-            return false;
-        }
-        if (getClass() != obj.getClass()) {
-            return false;
-        }
-        NameMetadataDescription other = (NameMetadataDescription) obj;
-        if (hasFirstComponentHash != other.hasFirstComponentHash) {
+        if (obj == null || getClass() != obj.getClass()) {
             return false;
         }
-        if (rowParts == null) {
-            if (other.rowParts != null) {
-                return false;
-            }
-        } else if (!rowParts.equals(other.rowParts)) {
+
+        NameMetadataDescription that = (NameMetadataDescription) obj;
+
+        if (numberOfComponentsHashed != that.numberOfComponentsHashed) {
             return false;
         }
-        return true;
+        return rowParts != null ? rowParts.equals(that.rowParts) : that.rowParts == null;
+    }
+
+    @Override
+    public int hashCode() {
```

Comment:
nit: `hashCode` should live above equals as before, makes the diff smaller and is the general practice.